// @pretzel-aux/common/src/Components/LikedTracksList/LikedTracksList.tsx
import * as React from 'react';
import { compose, graphql, GraphqlQueryControls } from 'react-apollo';
import { loader } from 'graphql.macro';
import { LikedTracksQuery } from './schema';
import { TrackList } from '../../Styled';
const likedTracksQuery = loader('./liked-tracks.graphql');

interface GraphqlProps {
  data: GraphqlQueryControls & LikedTracksQuery;
}

type Props = GraphqlProps;
interface State {
  loadingMore: boolean;
}

export class LikedTracksListPresentation extends React.Component<Props, State> {
  state: State = {
    loadingMore: false,
  };

  public render() {
    const tracks = this.props.data.likedTracks && this.props.data.likedTracks.edges.map(edge => edge.node);
    return (
      <TrackList
        title="Likes"
        tracks={tracks}
        hasMore={this.props.data.likedTracks && this.props.data.likedTracks.pageInfo.hasNextPage}
        onMore={this.onMore}
        loading={this.props.data.loading}
        loadingMore={this.state.loadingMore}
      />
    );
  }

  private onMore = () => {
    const after = this.props.data.likedTracks.pageInfo.endCursor || '';
    console.debug('Loading More', {
      type: 'track',
      currentLength: this.props.data.likedTracks.edges.length,
      after,
    });
    this.setState({ loadingMore: true });
    this.props.data.fetchMore({
      variables: { after },
      updateQuery: (previousQueryResult: LikedTracksQuery, { fetchMoreResult }) => {
        this.setState({ loadingMore: false });
        if (!fetchMoreResult) {
          return previousQueryResult;
        }
        return {
          likedTracks: Object.assign({}, fetchMoreResult.likedTracks, {
            edges: [...previousQueryResult.likedTracks.edges, ...fetchMoreResult.likedTracks.edges],
          }),
        };
      },
    });
  };
}

export const LikedTracksList = compose(
  graphql(likedTracksQuery, {
    options: {
      variables: {
        after: '',
      },
      fetchPolicy: 'cache-and-network',
    },
  }),
)(LikedTracksListPresentation);
