// @pretzel-aux/common/src/Components/Settings/Components/CheckboxOption/Styles.tsx
import styled from 'styled-components/macro';
import { PretzelColors } from '../../../../Styled';

export const ExternalWrapper = styled.div``;

export const Wrapper = styled.div`
  color: ${PretzelColors.baseWhite};
  font-size: 14px;
  padding: 10px;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  display: flex;
`;

export const StyledText = styled.div`
  margin-left: 6px;
  cursor: ${(props: { disabled?: boolean }) => (props.disabled ? 'not-allowed' : 'pointer')};
  -webkit-app-region: no-drag !important;
  color: ${(props: { disabled?: boolean }) => (props.disabled ? PretzelColors.baseMuted : PretzelColors.baseWhite)};
`;

export const CheckboxGroup = styled.div``;

export const StyledInput = styled.input`
  display: none;
`;

export interface StyledCheckboxProperties {
  disabled?: boolean;
  checked?: boolean;
}
export const StyledCheckbox = styled('label')<StyledCheckboxProperties>`
  display: block;
  width: 16px;
  height: 16px;
  border: 1px solid ${PretzelColors.baseBackground4};
  background: ${PretzelColors.baseBackground2};
  cursor: ${(props: { disabled?: boolean }) => (props.disabled ? 'not-allowed' : 'pointer')};
  margin-right: 3px;
  position: relative;
  &:after {
    content: '';
    width: 9px;
    height: 5px;
    position: absolute;
    top: 4px;
    left: 3px;
    border: 3px solid ${props => (props.disabled ? PretzelColors.baseMuted : PretzelColors.baseWhite)};
    border-top: none;
    border-right: none;
    background: transparent;
    opacity: ${(props: { checked?: boolean }) => (props.checked ? 1 : 0)};
    -webkit-transform: rotate(-45deg);
    transform: rotate(-45deg);
  }
  &:hover {
    &:after {
      opacity: ${(props: { checked?: boolean; disabled?: boolean }) =>
        props.disabled ? (props.checked ? '1' : '0') : '0.3'};
    }
  }
`;
