// @pretzel-aux/common/src/Core/Player/Components/PlayerControls/Components/AddToPlaylist/types/AddToPlaylistsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: AddToPlaylistsListQuery
// ====================================================

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_artists[];
  album: AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_album | null;
  self: AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node_self;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges {
  __typename: "TrackEdge";
  /**
   * The item at the end of the edge.
   */
  node: AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges_node | null;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node_tracks {
  __typename: "TrackConnection";
  /**
   * A list of edges.
   */
  edges: (AddToPlaylistsListQuery_userPlaylists_edges_node_tracks_edges | null)[] | null;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges_node {
  __typename: "Playlist";
  id: string;
  name: string;
  description: string;
  tracks: AddToPlaylistsListQuery_userPlaylists_edges_node_tracks;
}

export interface AddToPlaylistsListQuery_userPlaylists_edges {
  __typename: "PlaylistEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: AddToPlaylistsListQuery_userPlaylists_edges_node | null;
}

export interface AddToPlaylistsListQuery_userPlaylists_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface AddToPlaylistsListQuery_userPlaylists {
  __typename: "PlaylistConnection";
  /**
   * A list of edges.
   */
  edges: (AddToPlaylistsListQuery_userPlaylists_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: AddToPlaylistsListQuery_userPlaylists_pageInfo;
}

export interface AddToPlaylistsListQuery {
  /**
   * Return user's playlists
   */
  userPlaylists: AddToPlaylistsListQuery_userPlaylists;
}

export interface AddToPlaylistsListQueryVariables {
  after?: string | null;
}
