// @pretzel-aux/common/src/Core/Player/Components/PlayerControls/Components/AddToPlaylist/types/PlayerControlsNewPlaylist.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PlaylistAttributes } from "./../../../../../../GraphQL/schema";

// ====================================================
// GraphQL mutation operation: PlayerControlsNewPlaylist
// ====================================================

export interface PlayerControlsNewPlaylist_createPlaylist_playlist_tracks_edges_node {
  __typename: "Track";
  id: string;
}

export interface PlayerControlsNewPlaylist_createPlaylist_playlist_tracks_edges {
  __typename: "TrackEdge";
  /**
   * The item at the end of the edge.
   */
  node: PlayerControlsNewPlaylist_createPlaylist_playlist_tracks_edges_node | null;
}

export interface PlayerControlsNewPlaylist_createPlaylist_playlist_tracks {
  __typename: "TrackConnection";
  /**
   * A list of edges.
   */
  edges: (PlayerControlsNewPlaylist_createPlaylist_playlist_tracks_edges | null)[] | null;
}

export interface PlayerControlsNewPlaylist_createPlaylist_playlist {
  __typename: "Playlist";
  id: string;
  name: string;
  description: string;
  tracks: PlayerControlsNewPlaylist_createPlaylist_playlist_tracks;
}

export interface PlayerControlsNewPlaylist_createPlaylist {
  __typename: "CreatePlaylistPayload";
  /**
   * The playlist, null if create failed
   */
  playlist: PlayerControlsNewPlaylist_createPlaylist_playlist | null;
  /**
   * Any errors, for example if we are unable to create a Playlist.
   */
  errors: string[];
}

export interface PlayerControlsNewPlaylist {
  /**
   * Create a playlist for the current user.
   */
  createPlaylist: PlayerControlsNewPlaylist_createPlaylist;
}

export interface PlayerControlsNewPlaylistVariables {
  attributes: PlaylistAttributes;
}
