// @pretzel-aux/common/src/Components/StationsList/Components/StationCard.tsx
import * as React from 'react';
import { Gradient } from './Styles';
import { flowRight as compose } from 'lodash';
import { Box, Text } from 'grommet';
import { PlayerContext, withPlayerContext } from '../../../Core/Player/PlayerContext';
import { Icon, LibraryIcons, MusicCard, Night, PretzelColors, Square } from '../../../Styled';
import { minutesOrHours } from '../../../Util';
import { StationsListQuery_stations_edges_node } from '../types/StationsListQuery';
import { StationArt } from '../../../Styled/StationArt';

type PropsFromPlayer = Pick<PlayerContext, 'playSegment' | 'activeSegment'>;

interface PublicProps {
  station: StationsListQuery_stations_edges_node;
}

type Props = PublicProps & PropsFromPlayer;

export class StationCardPresentation extends React.Component<Props> {
  public render() {
    const { station, activeSegment } = this.props;
    const isPlaying = activeSegment && activeSegment.id === station.id;
    const isLiked = station.id === 'LIKED';
    const art = <StationArt color1={station.color1} color2={station.color2} theIcon={station.icon} />;
    const text = (
      <>
        <Text color={Night.textTitles}>{station.name}</Text>
        <Text size="xsmall">{station.description}</Text>
        <Box flex="grow" justify="end">
          {isLiked ? null : (
            <Text size="10px" color={Night.textMuted} margin={{ top: '16px' }} align-self="end">
              {station.stats.totalCount} Songs&nbsp;&nbsp;&nbsp;•&nbsp;&nbsp;&nbsp;
              {minutesOrHours(station.stats.duration)}
            </Text>
          )}
        </Box>
      </>
    );
    return <MusicCard art={art} text={text} active={isPlaying} onClick={this.handleClick} />;
  }
  private handleClick = () => {
    this.props.playSegment({ id: this.props.station.id });
  };
}

function mapContextToProps(c: PlayerContext): PropsFromPlayer {
  return {
    activeSegment: c.activeSegment,
    playSegment: c.playSegment,
  };
}

export const StationCard = compose(withPlayerContext(mapContextToProps))(StationCardPresentation);
