// @pretzel-aux/common/src/Core/Alerts/Files.ts
const AUDIO = process.env.PUBLIC_URL + '/audio/';

export interface Alert {
  id: string;
  name: string;
  src: string;
}

export const ALERT_1: Alert = {
  id: 'alert1',
  name: 'Alert 1',
  src: `${AUDIO}/alert1.mp3`,
};
export const ALERT_2: Alert = {
  id: 'alert2',
  name: 'Alert 2',
  src: `${AUDIO}/alert2.mp3`,
};
export const ALERT_3: Alert = {
  id: 'alert3',
  name: 'Alert 3',
  src: `${AUDIO}/alert3.mp3`,
};
export const ALERT_4: Alert = {
  id: 'alert4',
  name: 'Alert 4',
  src: `${AUDIO}/alert4.mp3`,
};
export const ALERT_5: Alert = {
  id: 'alert5',
  name: 'Alert 5',
  src: `${AUDIO}/alert5.mp3`,
};
export const ALERT_6: Alert = {
  id: 'alert6',
  name: 'Alert 6',
  src: `${AUDIO}/alert6.mp3`,
};
export const ALERT_7: Alert = {
  id: 'alert7',
  name: 'Alert 7',
  src: `${AUDIO}/alert7.mp3`,
};
export const ALERT_8: Alert = {
  id: 'alert8',
  name: 'Alert 8',
  src: `${AUDIO}/alert8.mp3`,
};
export const ALERT_9: Alert = {
  id: 'alert9',
  name: 'Alert 9',
  src: `${AUDIO}/alert9.mp3`,
};
export const ALERT_10: Alert = {
  id: 'alert10',
  name: 'Alert 10',
  src: `${AUDIO}/alert10.mp3`,
};
export const ALERT_11: Alert = {
  id: 'alert11',
  name: 'Alert 11',
  src: `${AUDIO}/alert11.mp3`,
};

export const Alerts: { [name: string]: Alert } = {
  [ALERT_1.id]: ALERT_1,
  [ALERT_2.id]: ALERT_2,
  [ALERT_3.id]: ALERT_3,
  [ALERT_4.id]: ALERT_4,
  [ALERT_5.id]: ALERT_5,
  [ALERT_6.id]: ALERT_6,
  [ALERT_7.id]: ALERT_7,
  [ALERT_8.id]: ALERT_8,
  [ALERT_9.id]: ALERT_9,
  [ALERT_10.id]: ALERT_10,
  [ALERT_11.id]: ALERT_11,
};
