// @pretzel-aux/common/src/Components/LabelsList/types/LabelsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: LabelsListQuery
// ====================================================

export interface LabelsListQuery_labels_edges_node_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface LabelsListQuery_labels_edges_node {
  __typename: "Label";
  id: string;
  artworkGuid: string;
  /**
   * The segment name
   */
  name: string;
  /**
   * Stats for this segment
   */
  stats: LabelsListQuery_labels_edges_node_stats;
}

export interface LabelsListQuery_labels_edges {
  __typename: "LabelEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: LabelsListQuery_labels_edges_node | null;
}

export interface LabelsListQuery_labels_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface LabelsListQuery_labels {
  __typename: "LabelConnection";
  /**
   * A list of edges.
   */
  edges: (LabelsListQuery_labels_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: LabelsListQuery_labels_pageInfo;
}

export interface LabelsListQuery {
  /**
   * Return all labels
   */
  labels: LabelsListQuery_labels;
}

export interface LabelsListQueryVariables {
  after?: string | null;
}
