// @pretzel-aux/common/src/Core/GraphQL/schema.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum PlayTokenSourceKindEnum {
  ARTIST = "ARTIST",
  DEFAULT = "DEFAULT",
  IMMEDIATE = "IMMEDIATE",
  LABEL = "LABEL",
  MIX = "MIX",
  PLAYLIST = "PLAYLIST",
  QUEUED = "QUEUED",
  RELEASE = "RELEASE",
  REPLAY = "REPLAY",
  SONG_REQUEST = "SONG_REQUEST",
  STATION = "STATION",
}

/**
 * Attributes for creating and updating a playlist
 */
export interface PlaylistAttributes {
  name: string;
  description: string;
  share?: boolean | null;
  trackIds: string[];
}

/**
 * Settings attributes for a user account
 */
export interface SettingsAttributes {
  audioVolume?: number | null;
  youtubeSafe?: boolean | null;
  allowExplicit?: boolean | null;
  autoplayOnLoad?: boolean | null;
  instrumentalOnly?: boolean | null;
  dailyAffirmations?: boolean | null;
  notifyDelay?: number | null;
  notifyInChat?: boolean | null;
  notifyOnlyWhenLive?: boolean | null;
}

//==============================================================
// END Enums and Input Objects
//==============================================================
