// @pretzel-aux/common/src/Core/Player/Components/LargeAlbumArt/Styles.tsx
import styled from 'styled-components/macro';
import { PretzelColors } from '../../../../Styled';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export const ArtWrapper = styled.div`
  width: 250px;
  height: 250px;
  @media (max-width: 470px) {
    width: 100%;
    height: auto;
    min-height: 250px;
  }
  @media (max-height: 460px) {
    max-width: 200px;
    max-height: 200px;
    min-height: 0px;
  }
  position: relative;
`;

export const Art = styled.img`
  width: 250px;
  height: 250px;
  object-fit: cover;
  display: flex;
  align-self: center;
  @media (max-width: 470px) {
    width: 100%;
    height: auto;
    min-height: 250px;
  }
  @media (max-height: 460px) {
    max-width: 200px;
    max-height: 200px;
    min-height: 0px;
  }
`;

export const AddToPlaylistWrapper = styled.div`
  width: 34px;
  height: 34px;
  position: absolute;
  bottom: 8px;
  right: 8px;
  border-radius: 50%;
  background-color: ${PretzelColors.baseBackground2};
  opacity: 0.5;
  display: flex;
  justify-content: center;
  align-items: center;
  &:hover {
    opacity: 1;
  }
  cursor: pointer;
`;

export const AddIcon = styled.i`
  color: ${PretzelColors.baseWhite};
  font-size: 22px;
  padding-right: 2px;
`;

export const SongRequestOverlay = styled.div`
  height: 25px;
  background: rgba(27, 179, 164, 0.9);
  width: 100%;
  position: absolute;
  top: 0;
  right: 0;
  display: flex;
  justify-content: space-between;
  align-items: center;
  color: #ffffff;
  font-size: 14px;
  padding: 0 10px;
`;

export const SRName = styled.div`
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-weight: 700;
`;

export const SRAmount = styled.div`
  text-align: right;
  font-weight: 600;
  margin-left: 5px;
`;
