// @pretzel-aux/common/src/Components/PlaylistsList/Components/PlaylistCard.tsx
import * as React from 'react';
import { flowRight as compose } from 'lodash';
import { Box, Image, Text } from 'grommet';
import { ArtworkSize, minutesOrHours, toUrl } from '../../../Util';
import { PlayerContext, withPlayerContext } from '../../../Core/Player/PlayerContext';
import { MusicCard, Night } from '../../../Styled';
import { PlaylistsListQuery_playlists_edges_node } from '../types/PlaylistsListQuery';

type PropsFromPlayer = Pick<PlayerContext, 'playSegment' | 'activeSegment'>;

interface PublicProps {
  playlist: PlaylistsListQuery_playlists_edges_node;
}

type Props = PublicProps & PropsFromPlayer;

export class PlaylistCardPresentation extends React.Component<Props> {
  public render() {
    const { playlist, activeSegment } = this.props;
    const isPlaying = activeSegment && activeSegment.id === playlist.id;
    const art = <Image src={toUrl(playlist.artworkGuid, ArtworkSize.Medium)} width="100%" />;
    const text = (
      <>
        <Box>
          <Text color={Night.textTitles}>{playlist.name}</Text>
        </Box>
        <Text size="xsmall">{playlist.description}</Text>
        <Box flex="grow" justify="end">
          <Text size="10px" color={Night.textMuted} margin={{ top: '16px' }} align-self="end">
            {playlist.stats.totalCount} Songs&nbsp;&nbsp;&nbsp;•&nbsp;&nbsp;&nbsp;
            {minutesOrHours(playlist.stats.duration)}
          </Text>
        </Box>
      </>
    );
    return (
      <MusicCard
        art={art}
        text={text}
        active={isPlaying}
        linkTo={`/segment/${playlist.id}`}
        onClick={this.handleClick}
      />
    );
    // return <MusicCard art={art} text={text} active={isPlaying} onClick={this.handleClick} linkTo={`/segment/${album.id}`} />;
  }
  private handleClick = () => {
    this.props.playSegment({ id: this.props.playlist.id });
  };
}

function mapContextToProps(c: PlayerContext): PropsFromPlayer {
  return {
    activeSegment: c.activeSegment,
    playSegment: c.playSegment,
  };
}

export const PlaylistCard = compose(withPlayerContext(mapContextToProps))(PlaylistCardPresentation);
