// @pretzel-aux/common/src/Core/Player/Components/PlayerControls/PlayerControls.tsx
import * as React from 'react';
import { withMutation } from '@apollo/client/react/hoc';
import { flowRight as compose } from 'lodash';
import { PositioningRoot, Wrapper } from './Styles';
import { ControlItem } from './Components/ControlItem/ControlItem';
import { VolumeControl } from './Components/VolumeControl/VolumeControl';
import { loader } from 'graphql.macro';
import {
  SetTrackBlacklistedInput,
  SetTrackBlacklistedMutation,
  SetTrackLikedInput,
  SetTrackLikedMutation,
} from './schema';
import { AnalyticsEventType } from '../../../Analytics';
import { Mutation } from '../../../GraphQL';
import { PlayerContext, withPlayerContext } from '../../PlayerContext';
import { PlayTokenFragment } from '../../Queries/types/PlayTokenFragment';
import { AddToPlaylist } from './Components/AddToPlaylist/AddToPlaylist';
import { PlatformContext, withPlatformContext } from '../../../Platform';
import { CompactOnly, LibraryOnly } from '../../../../Util/responsive';

const likeTrackMutation = loader('../../Queries/set-track-liked.graphql');
const blacklistTrackMutation = loader('../../Queries/set-track-blacklisted.graphql');

type PropsFromPlatform = Pick<PlatformContext, 'pretzelUser'>;

interface PropsFromContext {
  playToken: PlayTokenFragment | null;
  loadingNext: boolean;
  playing: boolean;
  muted: boolean;
  audioVolume: number;
  next: () => void;
  previous: () => void;
  togglePause: () => void;
  toggleMute: () => void;
  setVolume: (vol: number, instant?: boolean, faded?: boolean, visible?: boolean) => void;
  increaseVolume: () => void;
  decreaseVolume: () => void;
  sendAnalyticsEvent: (type: AnalyticsEventType, playToken?: PlayTokenFragment | null, source?: string) => void;
}

interface GraphQLProps {
  likeTrack: Mutation<SetTrackLikedInput, SetTrackLikedMutation>;
  blacklistTrack: Mutation<SetTrackBlacklistedInput, SetTrackBlacklistedMutation>;
}

type Props = PropsFromContext & GraphQLProps & PropsFromPlatform;

interface State {
  volumeBarVisible: boolean;
  loadingLike: boolean;
  loadingBlacklist: boolean;
  showAddToPlaylist: boolean;
}

class PlayerControlsPresentation extends React.Component<Props, State> {
  state = {
    volumeBarVisible: false,
    loadingLike: false,
    loadingBlacklist: false,
    showAddToPlaylist: false,
  };

  public render() {
    const { playToken } = this.props;
    const { premium } = this.props.pretzelUser;
    return (
      <Wrapper>
        <CompactOnly>
          <ControlItem
            isLoading={this.state.loadingLike}
            isDisabled={this.props.loadingNext}
            active={playToken && playToken.track.self.liked}
            name="thumbs_up"
            onClick={this.handleLike}
          />
          <ControlItem
            isLoading={this.state.loadingBlacklist}
            isDisabled={this.props.loadingNext}
            active={playToken && playToken.track.self.blacklisted}
            name="thumbs_down"
            onClick={this.handleBlacklist}
          />
        </CompactOnly>
        <LibraryOnly>
          <PositioningRoot onMouseLeave={this.handlePlaylistUnhover}>
            {this.state.showAddToPlaylist && <AddToPlaylist hide={this.handlePlaylistUnhover} />}
            <ControlItem
              isLoading={this.props.loadingNext}
              isDisabled={!premium}
              name="ListAdd"
              onClick={premium ? this.handleAddClick : undefined}
              style={{
                paddingTop: '5px',
              }}
              size={26}
            />
          </PositioningRoot>
          <ControlItem
            isLoading={this.props.loadingNext}
            isDisabled={false}
            size={26}
            name="RefreshArrow"
            onClick={this.handlePrevious}
          />
        </LibraryOnly>
        <ControlItem
          isLoading={this.props.loadingNext}
          play
          size={48}
          isDisabled={false}
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
          }}
          name={this.props.playing ? 'Pause' : 'PlayerPlay'}
          onClick={this.handleTogglePause}
        />
        <ControlItem isLoading={this.props.loadingNext} isDisabled={false} name="Next" onClick={this.handleNext} />
        <PositioningRoot
          style={{ margin: 6, width: 38 }}
          onMouseEnter={this.handleHoverable}
          onMouseLeave={this.handleHoverable}
        >
          <VolumeControl
            volumeBarVisible={this.state.volumeBarVisible}
            setVolume={this.props.setVolume}
            volume={this.props.audioVolume}
          />
          <ControlItem
            isLoading={false}
            isDisabled={false}
            style={{ margin: 0 }}
            name={
              this.props.muted
                ? 'Mute'
                : this.props.audioVolume > 0.4
                ? this.props.audioVolume > 0.8
                  ? 'Volume3'
                  : 'Volume2'
                : 'Volume'
            }
            onClick={this.handleMute}
          />
        </PositioningRoot>
      </Wrapper>
    );
  }

  public componentWillUnmount(): void {
    // @ts-ignore -- `removeEventListener` doesn't support passive:true
    document.removeEventListener('scroll', this.handleScroll, { passive: true });
  }

  private handleScroll = (event: WheelEvent) => {
    const direction = event.deltaY;
    if (direction) {
      if (direction < 0) {
        this.props.increaseVolume();
      } else if (direction > 0) {
        this.props.decreaseVolume();
      }
    }
  };

  private handleAddClick = () => {
    this.setState({ showAddToPlaylist: !this.state.showAddToPlaylist });
  };

  private handlePlaylistUnhover = () => {
    this.setState({ showAddToPlaylist: false });
  };

  private handleHoverable = () => {
    if (this.state.volumeBarVisible) {
      // @ts-ignore -- `removeEventListener` doesn't support passive:true
      document.removeEventListener('wheel', this.handleScroll, { passive: true });
    } else {
      document.addEventListener('wheel', this.handleScroll, { passive: true });
    }
    this.setState({ volumeBarVisible: !this.state.volumeBarVisible });
  };

  private handleLike = (event: React.MouseEvent<HTMLButtonElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    if (this.props.playToken) {
      this.setState({ loadingLike: true });
      this.props
        .likeTrack({
          variables: {
            trackId: this.props.playToken.track.id,
            value: !this.props.playToken.track.self.liked,
          },
        })
        .then(() => this.setState({ loadingLike: false }))
        .catch(() => this.setState({ loadingLike: false }));
      this.props.sendAnalyticsEvent(AnalyticsEventType.Like, this.props.playToken, 'playerControls');
    }
  };

  private handleBlacklist = (event: React.MouseEvent<HTMLElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    if (this.props.playToken) {
      this.setState({ loadingBlacklist: true });
      this.props
        .blacklistTrack({
          variables: {
            trackId: this.props.playToken.track.id,
            value: !this.props.playToken.track.self.blacklisted,
          },
        })
        .then(() => {
          this.setState({ loadingBlacklist: false });
          this.props.next();
        })
        .catch(() => this.setState({ loadingBlacklist: false }));
      this.props.sendAnalyticsEvent(AnalyticsEventType.Dislike, this.props.playToken, 'playerControls');
    }
  };

  private handleTogglePause = (event: React.MouseEvent<HTMLElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    this.props.togglePause();
  };

  private handleNext = (event: React.MouseEvent<HTMLElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    this.props.next();
  };

  private handlePrevious = (event: React.MouseEvent<HTMLElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    this.props.previous();
  };

  private handleMute = (event: React.MouseEvent<HTMLElement>) => {
    // @ts-ignore
    event.target.parentElement.blur();
    this.props.toggleMute();
  };
}

function mapContextToProps(c: PlayerContext): PropsFromContext {
  return {
    playToken: c.playToken,
    loadingNext: c.loadingNext,
    playing: c.playing,
    muted: c.muted,
    audioVolume: c.audioVolume,
    next: c.next,
    previous: c.previous,
    togglePause: c.togglePause,
    toggleMute: c.toggleMute,
    setVolume: c.setVolume,
    increaseVolume: c.increaseVolume,
    decreaseVolume: c.decreaseVolume,
    sendAnalyticsEvent: c.sendAnalyticsEvent,
  };
}

function mapPlatformToProps(c: PlatformContext): PropsFromPlatform {
  return {
    pretzelUser: c.pretzelUser,
  };
}

export const PlayerControls: React.ComponentClass = compose(
  withPlayerContext(mapContextToProps),
  withPlatformContext(mapPlatformToProps),
  withMutation(likeTrackMutation, { name: 'likeTrack' }),
  withMutation(blacklistTrackMutation, { name: 'blacklistTrack' }),
  // @ts-ignore Just need to move to hooks for these so that typing isn't so fucked up.
)(PlayerControlsPresentation);
