// @pretzel-aux/common/src/Components/ArtistsList/types/ArtistsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: ArtistsListQuery
// ====================================================

export interface ArtistsListQuery_artists_edges_node_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface ArtistsListQuery_artists_edges_node {
  __typename: "Artist";
  id: string;
  artworkGuid: string;
  /**
   * The segment name
   */
  name: string;
  /**
   * Stats for this segment
   */
  stats: ArtistsListQuery_artists_edges_node_stats;
}

export interface ArtistsListQuery_artists_edges {
  __typename: "ArtistEdge";
  /**
   * The item at the end of the edge.
   */
  node: ArtistsListQuery_artists_edges_node | null;
}

export interface ArtistsListQuery_artists_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface ArtistsListQuery_artists {
  __typename: "ArtistConnection";
  /**
   * A list of edges.
   */
  edges: (ArtistsListQuery_artists_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: ArtistsListQuery_artists_pageInfo;
}

export interface ArtistsListQuery {
  /**
   * Return all artists
   */
  artists: ArtistsListQuery_artists;
}

export interface ArtistsListQueryVariables {
  after?: string | null;
}
