// @pretzel-aux/common/src/Components/BlacklistedTracksList/BlacklistedTracksList.tsx
import * as React from 'react';
import { graphql, QueryControls } from '@apollo/client/react/hoc';
import { flowRight as compose } from 'lodash';
import { loader } from 'graphql.macro';
import { Box, Heading, Text } from 'grommet';
import { TrackList } from '../../Styled';
import { BlacklistedTracksQuery } from './types/BlacklistedTracksQuery';
const blacklistedTracksQuery = loader('./blacklisted-tracks.graphql');

interface GraphqlProps {
  data: QueryControls & BlacklistedTracksQuery;
}

type Props = GraphqlProps;
interface State {
  loadingMore: boolean;
}

export class BlacklistedTracksListPresentation extends React.Component<Props, State> {
  state: State = {
    loadingMore: false,
  };

  public render() {
    const tracks = this.props.data.blacklistedTracks && this.props.data.blacklistedTracks.edges.map(edge => edge.node);
    return this.props.data.loading || (tracks && tracks.length > 0) ? (
      <TrackList
        title="Dislikes"
        tracks={tracks}
        hasMore={this.props.data.blacklistedTracks && this.props.data.blacklistedTracks.pageInfo.hasNextPage}
        onMore={this.onMore}
        loading={this.props.data.loading}
        loadingMore={this.state.loadingMore}
      />
    ) : (
      <Box align="center" justify="center" height="100%" width="100%">
        <Heading level="2" margin={{ top: '20px', bottom: '5px' }}>
          You haven't disliked anything!
        </Heading>
        <Text>
          Clicking <i className="pretzel-icon-thumbs_down" /> will remove a song from being played.
        </Text>
      </Box>
    );
  }

  private onMore = () => {
    const after = this.props.data.blacklistedTracks.pageInfo.endCursor || '';
    console.debug('Loading More', {
      type: 'track',
      currentLength: this.props.data.blacklistedTracks.edges.length,
      after,
    });
    this.setState({ loadingMore: true });
    this.props.data.fetchMore({
      variables: { after },
      updateQuery: (previousQueryResult: BlacklistedTracksQuery, { fetchMoreResult }) => {
        this.setState({ loadingMore: false });
        if (!fetchMoreResult) {
          return previousQueryResult;
        }
        return {
          blacklistedTracks: Object.assign({}, fetchMoreResult.blacklistedTracks, {
            edges: [...previousQueryResult.blacklistedTracks.edges, ...fetchMoreResult.blacklistedTracks.edges],
          }),
        };
      },
    });
  };
}

export const BlacklistedTracksList = compose(
  graphql(blacklistedTracksQuery, {
    options: {
      variables: {
        after: '',
      },
      fetchPolicy: 'cache-and-network',
    },
  }),
)(BlacklistedTracksListPresentation);
