// @pretzel-aux/common/src/Components/AlbumsList/AlbumsList.tsx
import * as React from 'react';
import { compose, graphql, GraphqlQueryControls } from 'react-apollo';
import { loader } from 'graphql.macro';
import { AlbumsListQuery } from './schema';
import { AlbumCard } from './Components/AlbumCard';
import { MusicList } from '../../Styled';
const albumsQuery = loader('./albums.graphql');

interface GraphqlProps {
  data: GraphqlQueryControls & AlbumsListQuery;
}

type Props = GraphqlProps;

interface State {
  loadingMore: boolean;
}

export class AlbumsListPresentation extends React.Component<Props, State> {
  state: State = {
    loadingMore: false,
  };

  public render() {
    const albums = this.props.data.loading
      ? null
      : this.props.data.albums.edges.map(albumEdge => <AlbumCard key={albumEdge.node.id} album={albumEdge.node} />);
    const hasMore = this.props.data.albums && this.props.data.albums.pageInfo.hasNextPage;
    return (
      <MusicList
        title="Albums"
        items={albums}
        hasMore={hasMore}
        onMore={this.onMore}
        loading={this.state.loadingMore}
      />
    );
  }

  private onMore = () => {
    const after = this.props.data.albums.pageInfo.endCursor || '';
    console.debug('Loading More', {
      type: 'album',
      currentLength: this.props.data.albums.edges.length,
      after,
    });
    this.setState({ loadingMore: true });
    this.props.data.fetchMore({
      variables: { after },
      updateQuery: (previousQueryResult: AlbumsListQuery, { fetchMoreResult }) => {
        this.setState({ loadingMore: false });
        if (!fetchMoreResult) {
          return previousQueryResult;
        }
        return {
          albums: Object.assign({}, fetchMoreResult.albums, {
            edges: [...previousQueryResult.albums.edges, ...fetchMoreResult.albums.edges],
          }),
        };
      },
    });
  };
}

export const AlbumsList = compose(
  graphql(albumsQuery, {
    options: {
      variables: {
        after: '',
      },
    },
  }),
)(AlbumsListPresentation);
