// @pretzel-aux/common/src/Components/LeftNav/LeftNav.tsx
import { NavLink } from 'react-router-dom';
import * as React from 'react';
import { Box } from 'grommet';
import styled from 'styled-components';
import { CurrentUserMenu } from '../CurrentUser/CurrentUserMenu';
import { Icon } from '../../Styled';
import {
  HouseChimney1Alternate,
  LabelsIcon,
  ListBullets,
  MicrophoneKaraoke,
  PlaylistAlbumAlternate,
  RadioAntenna1,
} from '@pretzel-aux/assets/icons';

interface SvgIconInterface {
  left?: number;
  top?: number;
  right?: number;
}
const SvgIcon = styled('span')<SvgIconInterface>`
  display: inline-block;
  width: 18px;
  height: 18px;
  margin-right: ${({ right }) => (right ? `${right}px` : '4px')};
  margin-top: ${({ top }) => (top ? `${top}px` : 0)};
  margin-left: ${({ left }) => (left ? `${left}px` : 0)};
`;

const NavText = styled('span')`
  padding-left: 5px;
`;

export class LeftNav extends React.Component {
  public render() {
    return (
      <Box direction="column" justify="between" fill>
        <Box direction="column" pad={{ vertical: '16px' }} style={{ color: '#FFFFFF' }}>
          <NavLink to="/home" className="nav-link">
            {' '}
            <SvgIcon top={-5}>{<HouseChimney1Alternate />}</SvgIcon> <NavText>Home</NavText>
          </NavLink>
          <NavLink to="/stations" className="nav-link">
            {' '}
            <SvgIcon top={-5}>{<RadioAntenna1 />}</SvgIcon> <NavText>Stations</NavText>
          </NavLink>
          <NavLink to="/artists" className="nav-link">
            <SvgIcon>{<MicrophoneKaraoke />}</SvgIcon> <NavText>Artists</NavText>
          </NavLink>
          <NavLink to="/albums" className="nav-link">
            <SvgIcon>{<PlaylistAlbumAlternate />}</SvgIcon> <NavText>Albums</NavText>
          </NavLink>
          <NavLink to="/labels" className="nav-link">
            <SvgIcon left={-3} right={8}>
              {<LabelsIcon />}
            </SvgIcon>
            <NavText>Labels</NavText>
          </NavLink>
          <NavLink to="/playlists" className="nav-link">
            <SvgIcon left={-3} right={8}>
              {<ListBullets />}
            </SvgIcon>
            <NavText>Playlists</NavText>
          </NavLink>
          <Box height="30px" />
          <NavLink to="/likes" className="nav-link">
            <Icon name="thumbs_up" size={18} />
            <NavText>Likes</NavText>
          </NavLink>
          <NavLink to="/dislikes" className="nav-link">
            <Icon name="thumbs_down" size={18} />
            <NavText>Dislikes</NavText>
          </NavLink>
        </Box>
        <CurrentUserMenu />
      </Box>
    );
  }
}
