// @pretzel-aux/common/src/Components/SongRequests/types/SongRequestQueueQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SongRequestQueueQuery
// ====================================================

export interface SongRequestQueueQuery_currentUser_songRequestQueue_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface SongRequestQueueQuery_currentUser_songRequestQueue_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface SongRequestQueueQuery_currentUser_songRequestQueue_track {
  __typename: "Track";
  id: string;
  artworkGuid: string;
  title: string;
  artists: SongRequestQueueQuery_currentUser_songRequestQueue_track_artists[];
}

export interface SongRequestQueueQuery_currentUser_songRequestQueue {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: SongRequestQueueQuery_currentUser_songRequestQueue_requesters[];
  track: SongRequestQueueQuery_currentUser_songRequestQueue_track;
}

export interface SongRequestQueueQuery_currentUser {
  __typename: "User";
  id: string;
  songRequestsEnabled: boolean;
  songRequestQueue: SongRequestQueueQuery_currentUser_songRequestQueue[] | null;
}

export interface SongRequestQueueQuery {
  /**
   * The current user
   */
  currentUser: SongRequestQueueQuery_currentUser | null;
}
