// @pretzel-aux/common/src/Components/ArtistsList/ArtistsList.tsx
import * as React from 'react';
import { compose, graphql, GraphqlQueryControls } from 'react-apollo';
import { loader } from 'graphql.macro';
import { ArtistsListQuery } from './schema';
import { ArtistCard } from './Components/ArtistCard';
import { MusicList } from '../../Styled';
const artistsQuery = loader('./artists.graphql');

interface GraphqlProps {
  data: GraphqlQueryControls & ArtistsListQuery;
}

type Props = GraphqlProps;

interface State {
  loadingMore: boolean;
}

export class ArtistsListPresentation extends React.Component<Props, State> {
  state: State = {
    loadingMore: false,
  };

  public render() {
    const artists = this.props.data.loading
      ? null
      : this.props.data.artists.edges.map(artistEdge => (
          <ArtistCard key={artistEdge.node.id} artist={artistEdge.node} />
        ));
    const hasMore = this.props.data.artists && this.props.data.artists.pageInfo.hasNextPage;
    return (
      <MusicList
        title="Artists"
        items={artists}
        hasMore={hasMore}
        onMore={this.onMore}
        loading={this.state.loadingMore}
        circle={true}
      />
    );
  }

  private onMore = () => {
    const after = this.props.data.artists.pageInfo.endCursor || '';
    console.debug('Loading More', {
      type: 'artist',
      currentLength: this.props.data.artists.edges.length,
      after,
    });
    this.setState({ loadingMore: true });
    this.props.data.fetchMore({
      variables: { after },
      updateQuery: (previousQueryResult: ArtistsListQuery, { fetchMoreResult }) => {
        this.setState({ loadingMore: false });
        if (!fetchMoreResult) {
          return previousQueryResult;
        }
        return {
          artists: Object.assign({}, fetchMoreResult.artists, {
            edges: [...previousQueryResult.artists.edges, ...fetchMoreResult.artists.edges],
          }),
        };
      },
    });
  };
}

export const ArtistsList = compose(
  graphql(artistsQuery, {
    options: {
      variables: {
        after: '',
      },
    },
  }),
)(ArtistsListPresentation);
