// @pretzel-aux/common/src/Components/AlbumsList/types/AlbumsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: AlbumsListQuery
// ====================================================

export interface AlbumsListQuery_albums_edges_node_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface AlbumsListQuery_albums_edges_node_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface AlbumsListQuery_albums_edges_node {
  __typename: "Album";
  id: string;
  artworkGuid: string;
  /**
   * The segment name
   */
  name: string;
  artists: AlbumsListQuery_albums_edges_node_artists[];
  /**
   * Stats for this segment
   */
  stats: AlbumsListQuery_albums_edges_node_stats;
}

export interface AlbumsListQuery_albums_edges {
  __typename: "AlbumEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: AlbumsListQuery_albums_edges_node | null;
}

export interface AlbumsListQuery_albums_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface AlbumsListQuery_albums {
  __typename: "AlbumConnection";
  /**
   * A list of edges.
   */
  edges: (AlbumsListQuery_albums_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: AlbumsListQuery_albums_pageInfo;
}

export interface AlbumsListQuery {
  /**
   * Return all albums
   */
  albums: AlbumsListQuery_albums;
}

export interface AlbumsListQueryVariables {
  after?: string | null;
}
