// @pretzel-aux/common/src/Components/StationsList/types/StationsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: StationsListQuery
// ====================================================

export interface StationsListQuery_stations_edges_node_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface StationsListQuery_stations_edges_node {
  __typename: "Station";
  id: string;
  /**
   * The segment name
   */
  name: string;
  description: string;
  icon: string;
  color1: string;
  color2: string;
  /**
   * Stats for this segment
   */
  stats: StationsListQuery_stations_edges_node_stats;
}

export interface StationsListQuery_stations_edges {
  __typename: "StationEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: StationsListQuery_stations_edges_node | null;
}

export interface StationsListQuery_stations_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface StationsListQuery_stations {
  __typename: "StationConnection";
  /**
   * A list of edges.
   */
  edges: (StationsListQuery_stations_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: StationsListQuery_stations_pageInfo;
}

export interface StationsListQuery {
  /**
   * Return all stations
   */
  stations: StationsListQuery_stations;
}

export interface StationsListQueryVariables {
  after?: string | null;
  first?: number | null;
}
