// @pretzel-aux/common/src/Components/Search/Components/AlbumHit.tsx
import * as React from 'react';
import { compose } from 'react-apollo';
import { Box, Image, Text } from 'grommet';
import { Hit } from 'react-instantsearch-core';

import { Album } from '../schema';
import { PlayerContext, withPlayerContext } from '../../../Core/Player/PlayerContext';
import { toUrl } from '../../../Util';
import { MusicCard, Night } from '../../../Styled';

interface PublicProps {
  hit: Hit<Album>;
}

type PropsFromPlayer = Pick<PlayerContext, 'playSegment'>;

type Props = PublicProps & PropsFromPlayer

export class AlbumHitImplementation extends React.Component<Props> {
  public render() {
    const { hit } = this.props
    const art = <Image src={toUrl(hit.artwork_guid)} width="100%" />
    const text = (
      <>
        <Box>
          <Text color={Night.textTitles}>{hit.title}</Text>
        </Box>
        <Text size="xsmall">{hit.artists}</Text>
      </>
    )
    return <MusicCard art={art} text={text} onClick={this.handleClick} />
    // return <MusicCard art={art} text={text} onClick={this.handleClick} linkTo={`/segment/${hit.guid}`} />
  }

  private handleClick = () => {
    this.props.playSegment({ id: this.props.hit.guid });
  }
}

function mapContextToProps(c: PlayerContext): PropsFromPlayer {
  return {
    playSegment: c.playSegment,
  };
}
export const AlbumHit = compose(withPlayerContext(mapContextToProps))(AlbumHitImplementation);
