// @pretzel-aux/common/src/Core/Player/Components/PlayerControls/Components/ControlItem/Styles.tsx
import styled from 'styled-components/macro';
import { PretzelColors } from '../../../../../../Styled';

export const Wrapper = styled.div`
  position: relative;
`;

interface StyledButtonProps {
  active?: boolean | null;
  play?: boolean;
}

// For some reason, this component required an interface, either because it's a button, or because of `active`.
export const StyledButton = styled('button')<StyledButtonProps>`
  padding: 0;
  border: none;
  height: ${props => (props.play ? '48px' : '38px')};
  width: ${props => (props.play ? '48px' : '38px')};
  font-size: ${props => (props.play ? '36px' : '22px')};
  margin: 6px;
  text-align: center;
  background: transparent;
  color: ${props => (props.active ? PretzelColors.brandGreen2 : PretzelColors.baseWhite)};
  cursor: pointer;
  &:hover {
    color: ${PretzelColors.brandGreen2};
  }
  z-index: 3;
  position: relative;
`;

export const LoadingWrapper = styled.div`
  z-index: 10;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  justify-content: center;
  align-items: center;
`;
