// @pretzel-aux/common/src/Core/Player/Components/PlayerControls/Components/Settings/Styles.tsx
import styled from 'styled-components/macro';
import { PretzelColors } from '../../../../../../Styled';

export const Wrapper = styled.div`
  height: 74px;
  width: 100%;
  ${'' /* font-size: 12px; */}
  display: flex;
  flex-direction: column;
  align-items: center;
  position: absolute;
  bottom: 12px;
  background: ${PretzelColors.baseBackground};
  z-index: 7;
`;

export const Handle = styled.div`
  width: 100%;
  height: 12px;
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const OptionsWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  margin-top: 10px;
`;

export const OptionWrapper = styled.div`
  margin: 0 2.5px;
  cursor: pointer;

  ${({ disabled }: { disabled: boolean }) =>
    disabled
      ? `
    opacity: 0.6;
    pointer-events: none;
  `
      : ''}
`;

export const OptionIcon = styled.i`
  color: ${({ active }: { active: boolean }) => (active ? PretzelColors.brandGreen2 : PretzelColors.baseMuted)};
  font-size: 32px;
  height: 32px;
  cursor: pointer;
  &:hover {
    color: ${({ active }: { active: boolean }) => (active ? PretzelColors.brandGreen2 : PretzelColors.baseWhite)};
  }
`;
