// @pretzel-aux/common/src/Components/Search/types/SaturateTracksQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SaturateTracksQuery
// ====================================================

export interface SaturateTracksQuery_saturateTracks_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface SaturateTracksQuery_saturateTracks_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface SaturateTracksQuery_saturateTracks_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface SaturateTracksQuery_saturateTracks {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: SaturateTracksQuery_saturateTracks_artists[];
  album: SaturateTracksQuery_saturateTracks_album | null;
  self: SaturateTracksQuery_saturateTracks_self;
}

export interface SaturateTracksQuery {
  /**
   * Returns an array of tracks from a batch of guids
   */
  saturateTracks: SaturateTracksQuery_saturateTracks[];
}

export interface SaturateTracksQueryVariables {
  guids: string;
}
