// @pretzel-aux/common/src/Components/Search/types/SearchQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: SearchQuery
// ====================================================

export interface SearchQuery_search_tracks_edges_node_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface SearchQuery_search_tracks_edges_node_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface SearchQuery_search_tracks_edges_node_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface SearchQuery_search_tracks_edges_node {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: SearchQuery_search_tracks_edges_node_artists[];
  album: SearchQuery_search_tracks_edges_node_album | null;
  self: SearchQuery_search_tracks_edges_node_self;
}

export interface SearchQuery_search_tracks_edges {
  __typename: "TrackEdge";
  /**
   * The item at the end of the edge.
   */
  node: SearchQuery_search_tracks_edges_node | null;
}

export interface SearchQuery_search_tracks {
  __typename: "TrackConnection";
  /**
   * A list of edges.
   */
  edges: (SearchQuery_search_tracks_edges | null)[] | null;
}

export interface SearchQuery_search {
  __typename: "SearchResult";
  id: string;
  tracks: SearchQuery_search_tracks;
}

export interface SearchQuery {
  /**
   * Hyrdrate Data from Algolia Ids
   */
  search: SearchQuery_search | null;
}

export interface SearchQueryVariables {
  guids: string;
}
