// @pretzel-aux/common/src/Components/Settings/Sections/Playback.tsx
import * as React from 'react';
import { flowRight as compose } from 'lodash';
import { SettingsSection } from '../Components/SettingsSection';
import { CheckboxOption } from '../Components/CheckboxOption/CheckboxOption';
import { SettingsContext, withSettingsContext } from '../../../Core/Settings/Settings';

interface PublicProps {}
interface PropsFromContext {
  allowExplicit: boolean;
  instrumentalOnly: boolean;
  youtubeSafe: boolean;

  setAllowExplicit: (value: boolean) => void;
  setInstrumentalOnly: (value: boolean) => void;
  setYoutubeSafe: (value: boolean) => void;
}

type Props = PublicProps & PropsFromContext;

interface State {}

export class PlaybackPresentation extends React.Component<Props, State> {
  state = {};

  public render() {
    return (
      <SettingsSection title="Playback">
        <CheckboxOption
          label="Instrumental Only"
          onChange={this.onInstrumentalOnlyChanged}
          value={this.props.instrumentalOnly}
        />
        <CheckboxOption label="Allow Mature" onChange={this.onAllowExplicitChanged} value={this.props.allowExplicit} />
        <CheckboxOption label="YouTube Safe Mode" onChange={this.onYoutubeSafeChanged} value={this.props.youtubeSafe} />
      </SettingsSection>
    );
  }

  private onInstrumentalOnlyChanged = (value: boolean) => {
    this.props.setInstrumentalOnly(value);
  };

  private onAllowExplicitChanged = (value: boolean) => {
    this.props.setAllowExplicit(value);
  };

  private onYoutubeSafeChanged = (value: boolean) => {
    this.props.setYoutubeSafe(value);
  };
}

function mapContextToProps(c: SettingsContext): PropsFromContext {
  return {
    allowExplicit: c.allowExplicit,
    instrumentalOnly: c.instrumentalOnly,
    setAllowExplicit: c.setAllowExplicit,
    setInstrumentalOnly: c.setInstrumentalOnly,
    setYoutubeSafe: c.setYoutubeSafe,
    youtubeSafe: c.youtubeSafe,
  };
}

export const Playback = compose(withSettingsContext(mapContextToProps))(PlaybackPresentation);
