// @pretzel-aux/common/src/Components/PlaylistsList/types/UserPlaylistsListQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: UserPlaylistsListQuery
// ====================================================

export interface UserPlaylistsListQuery_userPlaylists_edges_node_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface UserPlaylistsListQuery_userPlaylists_edges_node {
  __typename: "Playlist";
  id: string;
  name: string;
  artworkGuid: string;
  description: string;
  /**
   * Stats for this segment
   */
  stats: UserPlaylistsListQuery_userPlaylists_edges_node_stats;
}

export interface UserPlaylistsListQuery_userPlaylists_edges {
  __typename: "PlaylistEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: UserPlaylistsListQuery_userPlaylists_edges_node | null;
}

export interface UserPlaylistsListQuery_userPlaylists_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface UserPlaylistsListQuery_userPlaylists {
  __typename: "PlaylistConnection";
  /**
   * A list of edges.
   */
  edges: (UserPlaylistsListQuery_userPlaylists_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: UserPlaylistsListQuery_userPlaylists_pageInfo;
}

export interface UserPlaylistsListQuery {
  /**
   * Return user's playlists
   */
  userPlaylists: UserPlaylistsListQuery_userPlaylists;
}

export interface UserPlaylistsListQueryVariables {
  after?: string | null;
}
