// @pretzel-aux/common/src/Styled/MenuList.tsx
import * as React from 'react';
import { Box, Button } from 'grommet';
import { ClipLoader } from 'halogenium';
import { PretzelColors } from './PretzelColors';

interface PublicProps<Thing> {
  title: string;
  items: Thing[] | null;
  hasMore: boolean;
  onMore: () => void;
  loading: boolean;
  circle?: boolean;
}

export class MenuList<Thing> extends React.Component<PublicProps<Thing>> {
  public render() {
    const items = this.props.items === null ? this.renderPlaceholders() : this.props.items;
    return (
      <div style={{ overflow: 'auto' }}>
        <Box>{items}</Box>
        <Box justify="center" align="center">
          {this.props.hasMore && (
            <Button size="small" color="control" onClick={this.props.onMore} disabled={this.props.loading}>
              Load More
            </Button>
          )}
          {this.props.loading && <ClipLoader color={PretzelColors.brandGreen} size="16px" />}
        </Box>
      </div>
    );
  }

  private renderPlaceholders() {
    return [1, 2, 3].map(i => <Box key={i} />);
  }
}
