// @pretzel-aux/common/src/Styled/SlobsColors.ts
export const teal = '#31c3a2';
export const purple = '#8736E0';
export const red = '#F85640';
export const yellowLight = '#FFF9E5';
export const yellowDark = '#CA9D00';
export const light1 = '#FFFFFF';
export const light2 = '#F5F8FA';
export const light3 = '#E3E8EB';
export const light4 = '#BDC2C4';
export const light5 = '#91979A';
export const dark1 = '#050D12';
export const dark2 = '#09161D';
export const dark3 = '#17242D';
export const dark3point8 = '#1f2a31';
export const dark4 = '#2B383F';
export const dark4point5 = '37484F';
export const dark5 = '#4F5E65';

export interface SlobsTheme {
  actionButtons: string;
  newButtons: string;
  warningButtons: string;
  info1: string;
  info2: string;
  bgVideoBackground: string;
  icons: string;
  textTitles: string;
  bgSections: string;
  bgMain: string;
  bgSectionsAlt: string;
  bgSectionsAltHover: string;
  textDefaultButtons: string;
  textParagraph: string;
  textMuted: string;
}

const common = {
  actionButtons: teal,
  newButtons: purple,
  warningButtons: red,
  info1: yellowLight,
  info2: yellowDark,
  bgVideoBackground: dark1,
  icons: light5,
};

export const Light: SlobsTheme = {
  ...common,
  textTitles: dark2,
  bgSections: light2,
  bgMain: light1,
  bgSectionsAlt: light2,
  bgSectionsAltHover: light3,
  textDefaultButtons: light3,
  textParagraph: dark5,
  textMuted: dark5,
};

export const Night: SlobsTheme = {
  ...common,
  textTitles: light1,
  bgSections: dark2,
  bgMain: dark3,
  bgSectionsAlt: dark4,
  bgSectionsAltHover: dark3point8,
  textDefaultButtons: dark5,
  textParagraph: light4,
  textMuted: light5,
};
