// @pretzel-aux/common/src/Components/LikedTracksList/types/LikedTracksQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: LikedTracksQuery
// ====================================================

export interface LikedTracksQuery_likedTracks_edges_node_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface LikedTracksQuery_likedTracks_edges_node_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface LikedTracksQuery_likedTracks_edges_node_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface LikedTracksQuery_likedTracks_edges_node {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: LikedTracksQuery_likedTracks_edges_node_artists[];
  album: LikedTracksQuery_likedTracks_edges_node_album | null;
  self: LikedTracksQuery_likedTracks_edges_node_self;
}

export interface LikedTracksQuery_likedTracks_edges {
  __typename: "TrackEdge";
  /**
   * A cursor for use in pagination.
   */
  cursor: string;
  /**
   * The item at the end of the edge.
   */
  node: LikedTracksQuery_likedTracks_edges_node | null;
}

export interface LikedTracksQuery_likedTracks_pageInfo {
  __typename: "PageInfo";
  /**
   * When paginating forwards, the cursor to continue.
   */
  endCursor: string | null;
  /**
   * When paginating forwards, are there more items?
   */
  hasNextPage: boolean;
  /**
   * When paginating backwards, are there more items?
   */
  hasPreviousPage: boolean;
  /**
   * When paginating backwards, the cursor to continue.
   */
  startCursor: string | null;
}

export interface LikedTracksQuery_likedTracks {
  __typename: "TrackConnection";
  /**
   * A list of edges.
   */
  edges: (LikedTracksQuery_likedTracks_edges | null)[] | null;
  /**
   * Information to aid in pagination.
   */
  pageInfo: LikedTracksQuery_likedTracks_pageInfo;
}

export interface LikedTracksQuery {
  /**
   * Return all of this user's liked tracks
   */
  likedTracks: LikedTracksQuery_likedTracks;
}

export interface LikedTracksQueryVariables {
  after: string;
}
