// @pretzel-aux/common/src/Core/Player/Queries/types/CurrentPlayTokenQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PlayTokenSourceKindEnum } from "./../../../GraphQL/schema";

// ====================================================
// GraphQL query operation: CurrentPlayTokenQuery
// ====================================================

export interface CurrentPlayTokenQuery_currentUser_activeSegment {
  __typename: "Album" | "Artist" | "Label" | "Mix" | "Playlist" | "Station";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SegmentSourceDetails {
  __typename: "SegmentSourceDetails";
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_artists[];
  album: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_album | null;
  self: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track_self;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_requesters[];
  track: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest_track;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails {
  __typename: "SongRequestSourceDetails";
  songRequest: CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails_songRequest;
}

export type CurrentPlayTokenQuery_currentUser_currentToken_source_details = CurrentPlayTokenQuery_currentUser_currentToken_source_details_SegmentSourceDetails | CurrentPlayTokenQuery_currentUser_currentToken_source_details_SongRequestSourceDetails;

export interface CurrentPlayTokenQuery_currentUser_currentToken_source {
  __typename: "PlayTokenSource";
  guid: string | null;
  kind: PlayTokenSourceKindEnum;
  name: string | null;
  details: CurrentPlayTokenQuery_currentUser_currentToken_source_details | null;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_track_self_audioUrls {
  __typename: "AudioUrls";
  low: string | null;
  normal: string;
  high: string | null;
  veryHigh: string | null;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
  audioUrls: CurrentPlayTokenQuery_currentUser_currentToken_track_self_audioUrls;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: CurrentPlayTokenQuery_currentUser_currentToken_track_artists[];
  album: CurrentPlayTokenQuery_currentUser_currentToken_track_album | null;
  self: CurrentPlayTokenQuery_currentUser_currentToken_track_self;
}

export interface CurrentPlayTokenQuery_currentUser_currentToken {
  __typename: "PlayToken";
  id: string;
  source: CurrentPlayTokenQuery_currentUser_currentToken_source;
  track: CurrentPlayTokenQuery_currentUser_currentToken_track;
}

export interface CurrentPlayTokenQuery_currentUser {
  __typename: "User";
  id: string;
  activeSegment: CurrentPlayTokenQuery_currentUser_activeSegment;
  currentToken: CurrentPlayTokenQuery_currentUser_currentToken | null;
}

export interface CurrentPlayTokenQuery {
  /**
   * The current user
   */
  currentUser: CurrentPlayTokenQuery_currentUser | null;
}
