// @pretzel-aux/common/src/Core/Player/Queries/types/NextPlayToken.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PlayTokenSourceKindEnum } from "./../../../GraphQL/schema";

// ====================================================
// GraphQL mutation operation: NextPlayToken
// ====================================================

export interface NextPlayToken_nextPlayToken_user_activeSegment {
  __typename: "Album" | "Artist" | "Label" | "Mix" | "Playlist" | "Station";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SegmentSourceDetails {
  __typename: "SegmentSourceDetails";
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_artists[];
  album: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_album | null;
  self: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track_self;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_requesters[];
  track: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest_track;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails {
  __typename: "SongRequestSourceDetails";
  songRequest: NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails_songRequest;
}

export type NextPlayToken_nextPlayToken_user_currentToken_source_details = NextPlayToken_nextPlayToken_user_currentToken_source_details_SegmentSourceDetails | NextPlayToken_nextPlayToken_user_currentToken_source_details_SongRequestSourceDetails;

export interface NextPlayToken_nextPlayToken_user_currentToken_source {
  __typename: "PlayTokenSource";
  guid: string | null;
  kind: PlayTokenSourceKindEnum;
  name: string | null;
  details: NextPlayToken_nextPlayToken_user_currentToken_source_details | null;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_track_self_audioUrls {
  __typename: "AudioUrls";
  low: string | null;
  normal: string;
  high: string | null;
  veryHigh: string | null;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
  audioUrls: NextPlayToken_nextPlayToken_user_currentToken_track_self_audioUrls;
}

export interface NextPlayToken_nextPlayToken_user_currentToken_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: NextPlayToken_nextPlayToken_user_currentToken_track_artists[];
  album: NextPlayToken_nextPlayToken_user_currentToken_track_album | null;
  self: NextPlayToken_nextPlayToken_user_currentToken_track_self;
}

export interface NextPlayToken_nextPlayToken_user_currentToken {
  __typename: "PlayToken";
  id: string;
  source: NextPlayToken_nextPlayToken_user_currentToken_source;
  track: NextPlayToken_nextPlayToken_user_currentToken_track;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: NextPlayToken_nextPlayToken_user_songRequestQueue_track_artists[];
  album: NextPlayToken_nextPlayToken_user_songRequestQueue_track_album | null;
  self: NextPlayToken_nextPlayToken_user_songRequestQueue_track_self;
}

export interface NextPlayToken_nextPlayToken_user_songRequestQueue {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: NextPlayToken_nextPlayToken_user_songRequestQueue_requesters[];
  track: NextPlayToken_nextPlayToken_user_songRequestQueue_track;
}

export interface NextPlayToken_nextPlayToken_user {
  __typename: "User";
  id: string;
  activeSegment: NextPlayToken_nextPlayToken_user_activeSegment;
  currentToken: NextPlayToken_nextPlayToken_user_currentToken | null;
  songRequestQueue: NextPlayToken_nextPlayToken_user_songRequestQueue[] | null;
}

export interface NextPlayToken_nextPlayToken {
  __typename: "NextPlayTokenPayload";
  /**
   * The current user
   */
  user: NextPlayToken_nextPlayToken_user;
  /**
   * Any errors, for example if we are unable to generate a PlayToken because the user has filtered out all possible tracks.
   */
  errors: string[];
}

export interface NextPlayToken {
  /**
   * Obtain a PlayToken for current user. Triggers consume & chat notifications
   */
  nextPlayToken: NextPlayToken_nextPlayToken;
}

export interface NextPlayTokenVariables {
  trackId?: string | null;
}
