// @pretzel-aux/common/src/Core/Player/Queries/types/PlayTokenFragment.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PlayTokenSourceKindEnum } from "./../../../GraphQL/schema";

// ====================================================
// GraphQL fragment: PlayTokenFragment
// ====================================================

export interface PlayTokenFragment_source_details_SegmentSourceDetails {
  __typename: "SegmentSourceDetails";
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_artists[];
  album: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_album | null;
  self: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track_self;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_requesters[];
  track: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest_track;
}

export interface PlayTokenFragment_source_details_SongRequestSourceDetails {
  __typename: "SongRequestSourceDetails";
  songRequest: PlayTokenFragment_source_details_SongRequestSourceDetails_songRequest;
}

export type PlayTokenFragment_source_details = PlayTokenFragment_source_details_SegmentSourceDetails | PlayTokenFragment_source_details_SongRequestSourceDetails;

export interface PlayTokenFragment_source {
  __typename: "PlayTokenSource";
  guid: string | null;
  kind: PlayTokenSourceKindEnum;
  name: string | null;
  details: PlayTokenFragment_source_details | null;
}

export interface PlayTokenFragment_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface PlayTokenFragment_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface PlayTokenFragment_track_self_audioUrls {
  __typename: "AudioUrls";
  low: string | null;
  normal: string;
  high: string | null;
  veryHigh: string | null;
}

export interface PlayTokenFragment_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
  audioUrls: PlayTokenFragment_track_self_audioUrls;
}

export interface PlayTokenFragment_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: PlayTokenFragment_track_artists[];
  album: PlayTokenFragment_track_album | null;
  self: PlayTokenFragment_track_self;
}

export interface PlayTokenFragment {
  __typename: "PlayToken";
  id: string;
  source: PlayTokenFragment_source;
  track: PlayTokenFragment_track;
}
