// @pretzel-aux/common/src/Core/Player/Queries/types/PlayTokenTrackFragment.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: PlayTokenTrackFragment
// ====================================================

export interface PlayTokenTrackFragment_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface PlayTokenTrackFragment_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface PlayTokenTrackFragment_self_audioUrls {
  __typename: "AudioUrls";
  low: string | null;
  normal: string;
  high: string | null;
  veryHigh: string | null;
}

export interface PlayTokenTrackFragment_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
  audioUrls: PlayTokenTrackFragment_self_audioUrls;
}

export interface PlayTokenTrackFragment {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: PlayTokenTrackFragment_artists[];
  album: PlayTokenTrackFragment_album | null;
  self: PlayTokenTrackFragment_self;
}
