// @pretzel-aux/common/src/Core/Player/Queries/types/PlayTokenSourceSongRequestFragment.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: PlayTokenSourceSongRequestFragment
// ====================================================

export interface PlayTokenSourceSongRequestFragment_songRequest_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface PlayTokenSourceSongRequestFragment_songRequest_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface PlayTokenSourceSongRequestFragment_songRequest_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface PlayTokenSourceSongRequestFragment_songRequest_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface PlayTokenSourceSongRequestFragment_songRequest_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: PlayTokenSourceSongRequestFragment_songRequest_track_artists[];
  album: PlayTokenSourceSongRequestFragment_songRequest_track_album | null;
  self: PlayTokenSourceSongRequestFragment_songRequest_track_self;
}

export interface PlayTokenSourceSongRequestFragment_songRequest {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: PlayTokenSourceSongRequestFragment_songRequest_requesters[];
  track: PlayTokenSourceSongRequestFragment_songRequest_track;
}

export interface PlayTokenSourceSongRequestFragment {
  __typename: "SongRequestSourceDetails";
  songRequest: PlayTokenSourceSongRequestFragment_songRequest;
}
