// @pretzel-aux/common/src/Core/Player/Queries/types/SetActiveSegment.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: SetActiveSegment
// ====================================================

export interface SetActiveSegment_setActiveSegment_user_activeSegment {
  __typename: "Album" | "Artist" | "Label" | "Mix" | "Playlist" | "Station";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface SetActiveSegment_setActiveSegment_user {
  __typename: "User";
  id: string;
  activeSegment: SetActiveSegment_setActiveSegment_user_activeSegment;
}

export interface SetActiveSegment_setActiveSegment {
  __typename: "SetActiveSegmentPayload";
  user: SetActiveSegment_setActiveSegment_user;
}

export interface SetActiveSegment {
  /**
   * Change the current user's active segment.
   */
  setActiveSegment: SetActiveSegment_setActiveSegment;
}

export interface SetActiveSegmentVariables {
  id: string;
}
