// @pretzel-aux/common/src/Core/Player/Queries/types/SongRequestWithTrackFragment.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: SongRequestWithTrackFragment
// ====================================================

export interface SongRequestWithTrackFragment_requesters {
  __typename: "SongRequestRequester";
  bitTotal: number;
  displayName: string;
}

export interface SongRequestWithTrackFragment_track_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface SongRequestWithTrackFragment_track_album {
  __typename: "Album";
  id: string;
  title: string;
}

export interface SongRequestWithTrackFragment_track_self {
  __typename: "SelfTrack";
  liked: boolean;
  blacklisted: boolean;
}

export interface SongRequestWithTrackFragment_track {
  __typename: "Track";
  id: string;
  title: string;
  duration: number | null;
  artworkGuid: string;
  explicit: boolean | null;
  instrumental: boolean | null;
  youtubeSafe: boolean;
  artists: SongRequestWithTrackFragment_track_artists[];
  album: SongRequestWithTrackFragment_track_album | null;
  self: SongRequestWithTrackFragment_track_self;
}

export interface SongRequestWithTrackFragment {
  __typename: "SongRequest";
  bitTotal: number;
  id: string;
  leaderName: string;
  others: number;
  requesters: SongRequestWithTrackFragment_requesters[];
  track: SongRequestWithTrackFragment_track;
}
