// @pretzel-aux/common/src/Components/Home/types/HomeDataQuery.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: HomeDataQuery
// ====================================================

export interface HomeDataQuery_homeData_featured_artists {
  __typename: "Artist";
  id: string;
  /**
   * The segment name
   */
  name: string;
}

export interface HomeDataQuery_homeData_featured_stats {
  __typename: "SegmentStats";
  /**
   * The total number of tracks in the segment
   */
  totalCount: number;
  /**
   * The number of seconds in the segment
   */
  duration: number;
  /**
   * The popularity of a segment, larger == more popular
   */
  popularity: number;
}

export interface HomeDataQuery_homeData_featured {
  __typename: "Album";
  id: string;
  artworkGuid: string;
  /**
   * The segment name
   */
  name: string;
  artists: HomeDataQuery_homeData_featured_artists[];
  /**
   * Stats for this segment
   */
  stats: HomeDataQuery_homeData_featured_stats;
}

export interface HomeDataQuery_homeData_gettingStarted {
  __typename: "ExternalLink";
  name: string;
  url: string;
}

export interface HomeDataQuery_homeData_comingSoon {
  __typename: "ExternalLink";
  name: string;
  url: string;
}

export interface HomeDataQuery_homeData {
  __typename: "HomeData";
  /**
   * Title for the featured section
   */
  featuredTitle: string;
  /**
   * New Releases
   */
  featured: HomeDataQuery_homeData_featured[];
  /**
   * Getting Started URLs
   */
  gettingStarted: HomeDataQuery_homeData_gettingStarted[];
  /**
   * Coming Soon Features
   */
  comingSoon: HomeDataQuery_homeData_comingSoon[];
}

export interface HomeDataQuery {
  homeData: HomeDataQuery_homeData;
}
