// @pretzel-aux/common/src/Components/LabelsList/Components/LabelCard.tsx
import * as React from 'react';
import { flowRight as compose } from 'lodash';
import { Box, Image, Text } from 'grommet';
import { PlayerContext, withPlayerContext } from '../../../Core/Player/PlayerContext';
import { ArtworkSize, minutesOrHours, toUrl } from '../../../Util';
import { MusicCard, Night } from '../../../Styled';
import { LabelsListQuery_labels_edges_node } from '../types/LabelsListQuery';

type PropsFromPlayer = Pick<PlayerContext, 'playSegment' | 'activeSegment'>;

interface PublicProps {
  label: LabelsListQuery_labels_edges_node;
}

type Props = PublicProps & PropsFromPlayer;

export class LabelCardPresentation extends React.Component<Props> {
  public render() {
    const { label, activeSegment } = this.props;
    const isPlaying = activeSegment && activeSegment.id === label.id;
    const art = (
      <Image src={toUrl(label.artworkGuid, ArtworkSize.Medium)} width="100%" style={{ borderRadius: '50%' }} />
    );
    const text = (
      <>
        <Text color={Night.textTitles}>{label.name}</Text>
        <Box flex="grow" justify="end">
          <Text size="10px" color={Night.textMuted} margin={{ top: '16px' }} align-self="end">
            {label.stats.totalCount} Songs&nbsp;&nbsp;&nbsp;•&nbsp;&nbsp;&nbsp;{minutesOrHours(label.stats.duration)}
          </Text>
        </Box>
      </>
    );
    return (
      <MusicCard
        art={art}
        text={text}
        active={isPlaying}
        circle={true}
        onClick={this.handleClick}
        linkTo={`/segment/${label.id}`}
      />
    );
  }
  private handleClick = () => {
    this.props.playSegment({ id: this.props.label.id });
  };
}

function mapContextToProps(c: PlayerContext): PropsFromPlayer {
  return {
    activeSegment: c.activeSegment,
    playSegment: c.playSegment,
  };
}

export const LabelCard = compose(withPlayerContext(mapContextToProps))(LabelCardPresentation);
