// @pretzel-aux/common/src/Core/Settings/Queries/types/SetSettings.ts
/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { SettingsAttributes } from "./../../../GraphQL/schema";

// ====================================================
// GraphQL mutation operation: SetSettings
// ====================================================

export interface SetSettings_updateSettings_user_settings {
  __typename: "Settings";
  /**
   * Should explicit music be allowed.
   */
  allowExplicit: boolean;
  /**
   * Volume setting percentage, values between 0 and 1 are accepted.
   */
  audioVolume: number;
  /**
   * Should the player immediately start playing when loaded.
   */
  autoplayOnLoad: boolean;
  /**
   * Does this user want their heart warmed on a daily basis?
   */
  dailyAffirmations: boolean;
  /**
   * Should music selections contain no vocals.
   */
  instrumentalOnly: boolean;
  /**
   * Delay the front-end should wait of continuous playback of a track before sending a notify message to the server.
   */
  notifyDelay: number;
  /**
   * When a PlayToken is consumed using the notify API, should Pretzel send a message to the user's channel chat room.
   */
  notifyInChat: boolean;
  /**
   * Preference to only send chat messages when a user is broadcasting. Handled server side.
   */
  notifyOnlyWhenLive: boolean;
  /**
   * Whether or not the play token generator should limit selects to tracks safe for YouTube.
   */
  youtubeSafe: boolean;
}

export interface SetSettings_updateSettings_user {
  __typename: "User";
  id: string;
  settings: SetSettings_updateSettings_user_settings;
}

export interface SetSettings_updateSettings {
  __typename: "UpdateSettingsPayload";
  user: SetSettings_updateSettings_user;
  /**
   * Any errors when saving settings
   */
  errors: string[];
}

export interface SetSettings {
  /**
   * Update settings.
   */
  updateSettings: SetSettings_updateSettings;
}

export interface SetSettingsVariables {
  attributes: SettingsAttributes;
}
