// @pretzel-aux/common/src/Components/Settings/Sections/FileOutput/OutputFileSelector.tsx
import React from 'react';
import styled from 'styled-components';
import { PretzelColors } from '../../../../Styled';
import { Box, Text } from 'grommet';

interface Disableable {
  disabled?: boolean;
}

const FileButton = styled.button<Disableable>`
  color: ${props => (props.disabled ? PretzelColors.baseMuted : PretzelColors.baseWhite)};
  background: transparent;
  font-size: 18px;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  &:hover {
    color: ${props => (props.disabled ? PretzelColors.baseMuted : PretzelColors.brandGreen2)};
  }
`;

const FileInput = styled.input`
  background-color: #fff;
  border-color: #d9d9d9 #ccc #b3b3b3;
  border-radius: 4px;
  border: 1px solid #ccc;
  color: #333;
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  display: table;
  border-spacing: 0;
  border-collapse: separate;
  height: 36px;
  outline: none;
  overflow: hidden;
  position: relative;
  width: 100%;
  line-height: 34px;
  white-space: nowrap;
  font-size: 14px;
  padding: 0 10px;
`;
const InputWrapper = styled.div`
  width: 98%;
`;
const Icon = styled.i<Disableable>`
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
`;

interface OutputFileSelectorProperties {
  disabled: boolean;
  title: string;
  filePath: string;
  openSaveDialog: () => void;
  clearSavePath: () => void;
}

export const OutputFileSelector = (props: OutputFileSelectorProperties) => (
  <Box margin="10px">
    <Text size="small" color={props.disabled ? PretzelColors.baseMuted : PretzelColors.baseWhite}>
      {props.title}
    </Text>
    <Box direction="row">
      <InputWrapper>
        <FileInput
          disabled={props.disabled}
          type="text"
          readOnly
          value={props.filePath || ''}
          onClick={(e: React.MouseEvent<HTMLInputElement>) => {
            props.openSaveDialog();
          }}
        />
      </InputWrapper>
      <FileButton disabled={props.disabled} onClick={props.openSaveDialog} title="Edit">
        <Icon disabled={props.disabled} className="pretzel-icon-create_new" />
      </FileButton>
      <FileButton disabled={props.disabled} onClick={props.clearSavePath} title="Clear">
        <Icon disabled={props.disabled} className="pretzel-icon-close" />
      </FileButton>
    </Box>
  </Box>
);
